patterns/ffi-export.md

```diff
@@ -0,0 +1,180 @@
+
+Rust is well known for its iterators. When implementing one, the programmer makes a separate type with a bounded lifetime to its owner, and implements the `Iterator` trait.
+
+Here is how iteration would be done in Rust for `DBM`:
+
+```rust,ignore
+struct Dbm { ... }
+
+impl Dbm {
+    /* ... */
+    pub fn keys<'it>(&'it self) -> DbmKeysIter<'it> { ... }
+    /* ... */
+}
+
+struct DbmKeysIter<'it> {
+    owner: &'it Dbm,
+}
+
+impl<'it> Iterator for DbmKeysIter<'it> { ... }
+```
+
+This is clean, idiomatic, and safe. thanks to Rust's guarantees. However, consider what a straightforward API translation would look like:
+
+```rust,ignore
+#[no_mangle]
+pub extern "C" fn dbm_iter_new(owner: *const Dbm) -> *mut DbmKeysIter {
+    /* THIS API IS A BAD IDEA! For real applications, use object-based design instead. */
+}
+#[no_mangle]
+pub extern "C" fn dbm_iter_next(iter: *mut DbmKeysIter, key_out: *const datum) -> libc::c_int {
+    /* THIS API IS A BAD IDEA! For real applications, use object-based design instead. */
+}
+#[no_mangle]
+pub extern "C" fn dbm_iter_del(*mut DbmKeysIter) {
+    /* THIS API IS A BAD IDEA! For real applications, use object-based design instead. */
+}
+```    
+
+This API loses a key piece of information: the lifetime of the iterator must not exceed the lifetime of the `Dbm` object that owns it. A user of the library could which causes the iterator to outlive the data it is iterating on, resulting in reading uninitialized memory.
+
+For the C programmers out there, spot the bug:
+
+```C
+int count_key_sizes(DBM *db) {
+    /* DO NOT USE THIS FUNCTION. IT HAS A SUBTLE BUT SERIOUS BUG! */
+    datum key;
+    int len = 0;
+
+    if (!dbm_iter_new(db)) {
+        dbm_close(db);
+        return -1;
+    }
+
+    int l;
+    while ((l = dbm_iter_next(owner, &key)) >= 0) { // an error is indicated by -1
+        free(key.dptr);
+        len += key.dsize;
+        if (l == 0) { // end of the iterator
+            dbm_close(owner);
+        }
+    }
+    if l >= 0 {
+        return -1;
+    } else {
+        return len;
+    }
+}
+```
+
+This bug is a classic. Here's what happens when the iterator returns the end-of-iteration marker:
+
+1. The loop condition sets `l` to zero, and enters the loop because `0 >= 0`.
+1. The length is incremented, in this case by zero.
+1. The if statement is true, so the database is closed. There should be a break statement here.
+1. The loop condition executes again, causing a `next` call on the closed object.
+
+The worst part about this bug? If the Rust implementation was careful, this code will work most of the time! If the memory for the `Dbm` object is not immediately reused, an internal check will almost certainly return an error, resulting in the iterator returning a -1. But occasionally, it will be cause a segmentation fault, or even worse nonsensical memory corruption!
```

Comment:
```suggestion
The worst part about this bug? If the Rust implementation was careful, this code will work most of the time!
If the memory for the `Dbm` object is not immediately reused, an internal check will almost certainly return an error, resulting in the iterator returning a `-1` indicating an error.
But occasionally, it will cause a segmentation fault, or even worse nonsensical memory corruption!
```